### ISSUE TITLE
SPA won't run when using guarded browser functions inside +layout.ts/+page.ts

### ISSUE THREAD

**Issue by freb** [sveltejs/kit#8356]
### Describe the bug

The gist of the issue is that a properly configured SK SPA application that uses browser API code inside of a `+layout.ts/+page.ts` file breaks `npm run dev` and `npm run preview` even when the browser code is guarded by the `building` variable.

I've been running into this issue since updating to v1.0 of SK and I belive it has to do with this change: https://github.com/sveltejs/kit/discussions/7716

I have a settings store that is backed by localStorage that I import pretty much everywhere. When trying to build I get the following error:

```
file:///home/me/code/cas/web/.svelte-kit/output/server/chunks/settings.js:23
  const val = localStorage.getItem(settingsKey) || "{}";
              ^

ReferenceError: localStorage is not defined
    at localSettings (file:///home/me/code/cas/web/.svelte-kit/output/server/chunks/settings.js:23:15)
...
```

I figured the above referenced changes were the issue, so I guarded the localStorage references inside my settings store. This allowed me to successfully build with `npm run build`.

However, when I try to serve up the build with `npm run preview` or run `npm run dev`, I get the same build error. If SSR and prerender are disabled and adapter-static is properly configured for SPA, then to my knowledge the code should only run on the server during build, but obviously the code is either running outside of build or SK does not properly set the `building` variable when running dev or preview.

I've create a minimal repoduction of the issue in this repo:

https://github.com/freb/sk-spa-issue

There you'll be able to confirm that I'm setting the proper exports in the root layout and the adapter-static settings in svelte.config.js.

Interestingly, while creating the reproduction I discovered the issue does not manifest when importing my settings store into `+page.svelte`, but only into `+layout.ts` or `+page.ts`.

### Reproduction

<https://github.com/freb/sk-spa-issue>

### Logs

```Shell
file:///home/me/code/cas/web/.svelte-kit/output/server/chunks/settings.js:23
  const val = localStorage.getItem(settingsKey) || "{}";
              ^

ReferenceError: localStorage is not defined
    at localSettings (file:///home/me/code/cas/web/.svelte-kit/output/server/chunks/settings.js:23:15)
    at createSettings (file:///home/me/code/cas/web/.svelte-kit/output/server/chunks/settings.js:39:21)
    at file:///home/me/code/cas/web/.svelte-kit/output/server/chunks/settings.js:59:18
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)

Node.js v19.3.0
[vite-plugin-sveltekit-compile] Prerendering failed with code 1
error during build:
Error: Prerendering failed with code 1
    at ChildProcess.<anonymous> (file:///home/me/code/cas/web/node_modules/@sveltejs/kit/src/exports/vite/index.js:551:15)
    at ChildProcess.emit (node:events:513:28)
    at ChildProcess._handle.onexit (node:internal/child_process:293:12)
```

### System Info

```Shell
System:
    OS: Linux 6.1 Arch Linux
    CPU: (24) x64 AMD Ryzen Threadripper 1920X 12-Core Processor
    Memory: 52.73 GB / 62.65 GB
    Container: Yes
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 19.3.0 - /usr/bin/node
    Yarn: 1.22.19 - /usr/bin/yarn
    npm: 8.19.2 - /usr/bin/npm
  npmPackages:
    @sveltejs/adapter-auto: ^1.0.0 => 1.0.0 
    @sveltejs/adapter-static: ^1.0.0 => 1.0.0 
    @sveltejs/kit: ^1.0.0 => 1.0.1 
    svelte: ^3.54.0 => 3.55.0 
    vite: ^4.0.0 => 4.0.4
```

### Severity

blocking all usage of SvelteKit

**Comment by Rich-Harris** [comment]
You're importing a module (`src/stores/settings.ts`) that proactively accesses `localStorage` as soon as it's imported, whether that happens on the server or in the browser. `building` is irrelevant, the guard you need is this:

```diff
-import { building } from '$app/environment';
+import { browser } from '$app/environment';

// later...
-if (building) {
+if (!browser) {
  return;
}

**Comment by freb** [comment]
Thanks, Rich. 

I'm confident that will fix the issue, though I'm not sure why whether or not the app is building is irrelevant. Prior to the referenced changes (at least at some point before the changes referenced [here](https://github.com/sveltejs/kit/discussions/7716)) the old escape hatch for disabling SSR (from hooks.js) worked just fine with the `npm run dev` command which at the time used `svelte-kit`.

After the referenced changes I was under the impression that server-side code should only be running during build time or when SSR is enabled. My guess is that the Vite server is taking some liberties. I'd love to know if this is just how it is with the vite server or if there is something fundamental I'm not understanding about disabling server-side execution of code. My closest guess is that vite has to run the code for HMR, but again it was working with previous versions of SK without guards.

**Comment by Rich-Harris** [comment]
You're not disabling the server, you're just disabling SSR. Your `+layout.ts` and `+page.ts` files need to be imported in order to determine whether SSR is enabled for a given page, and since `settings.ts` is imported by those files, `localStorage` is accessed as soon as `handle` is invoked.

If you were only importing `settings.ts` from your `+layout.svelte` and `+page.svelte` files, the code wouldn't run on the server.

**Comment by freb** [comment]
But isn't the importing of `+layout.ts` and `+page.ts` only done during build time? I realize I'm not disabling the server, but shouldn't the function of the Vite server be to simply deliver the static files (particularly in preview mode) or to watch for changes and recompile (in dev mode). And if Vite server is watching and recompiling, shouldn't that set the build flag to true?

I'm just trying to understand when Vite server would be importing `+layout.ts` during a time it is not performing a build.

To take the clearest example, the Vite docs [claim](https://vitejs.dev/guide/static-deploy.html):

> The vite preview command will boot up a local static web server that serves the files from dist at http://localhost:4173. It's an easy way to check if the production build looks OK in your local environment. 

This is clearly not what it is doing since I get an error on my terminal from when Vite tries to run code (not a browser error). But if I serve the same build folder with Caddy, it works just fine.

**Comment by Rich-Harris** [comment]
> But isn't the importing of `+layout.ts` and `+page.ts` only done during build time?

No. It's done whenever the server responds to a request, whether that's done during `dev`, `preview` or prerendering/building (which in your case is disabled).

**Comment by freb** [comment]
After writing what I have below, I realized that using browser functionality outside of `+page/layout.svelte` files is probably not very common. And given how much code needs to be guarded, your solution is very workable. Though I do think my particular use case would be reasonably common in an SPA (i.e. get a session token from localStorage and check its validity in `+page/layout.js` before rendering the page).

I'm going to look into Vite a little more before I let this drop, but I'm going to close the issue. Thanks a bunch for a solution and helping me get at the root of the issue. I leave my first draft of this comment below for posterity:

---

I think we're talking past each other. I'm assuming a world where node.js is not running on the server because I'm developing an SPA application. SvelteKit seems to support SPA's with the options I've included in my example project. I've confirmed that with proper build guards for browser functions, I can build and deploy the app just fine.

The issue is that Vite does not seem to know that I am developing an SPA where no code is intended to run on the server. So using `vite dev` and `vite preview` do not work as expected, because they attempt to execute my code.

As you mentioned, I can further guard my code so that if it were to run on a node.js server it would not error, but that's not what someone building an SPA would expect.

It seems that either Vite itself or the way that SvelteKit integrates with Vite does not have a way of preventing project code from being executed (omitting the execution during build time which is documented and expected).

There seem to be three options:

1. Fix Vite or the SK/Vite combo so that it is possible to serve code without executing it outside of build time
2. Realize that Vite is going to execute your code no matter what and use guards based on `browser` instead of `building`
3. Realize that Vite is going to execute your code no matter what so change `package.json` to a different static web server for at least `npm run preview` and possibly `npm run dev` (since vite won't work anyway)
